@@ -1,82 +0,0 @@
From f34d45478594590a4b30fdff6968d258d5883919 Mon Sep 17 00:00:00 2001
From: James Haggerty <james.haggerty@morsemicro.com>
Date: Fri, 28 Jun 2024 09:28:52 +1000
Subject: [PATCH] ARM: dts: bcm2711: Add morse HAT gpios to gpio-line-names

---
 arch/arm/boot/dts/broadcom/bcm2710-rpi-3-b.dts       | 6 +++---
 arch/arm/boot/dts/broadcom/bcm2710-rpi-zero-2-w.dts  | 6 +++---
 arch/arm/boot/dts/broadcom/bcm2711-rpi-4-b.dts       | 6 +++---
 arch/arm/boot/dts/broadcom/bcm2711-rpi-cm4.dts       | 6 +++---
 arch/arm64/boot/dts/broadcom/bcm2712-rpi-5-b.dts     | 6 +++---
 arch/arm64/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi    | 6 +++---
 6 files changed, 18 insertions(+), 18 deletions(-)

--- a/arch/arm/boot/dts/broadcom/bcm2710-rpi-3-b.dts
+++ b/arch/arm/boot/dts/broadcom/bcm2710-rpi-3-b.dts
@@ -48,9 +48,9 @@
 			  "GPIO15",
 			  "GPIO16",
-			  "GPIO17",
+			  "MM_RESET",
 			  "GPIO18",
 			  "GPIO19",
 			  "GPIO20",
 			  "GPIO21",
 			  "GPIO22",
-			  "GPIO23",
-			  "GPIO24",
+			  "MM_WAKE",
+			  "MM_BUSY",
 			  "GPIO25",
--- a/arch/arm/boot/dts/broadcom/bcm2710-rpi-zero-2-w.dts
+++ b/arch/arm/boot/dts/broadcom/bcm2710-rpi-zero-2-w.dts
@@ -46,9 +46,9 @@
 			  "GPIO15",
 			  "GPIO16",
-			  "GPIO17",
+			  "MM_RESET",
 			  "GPIO18",
 			  "GPIO19",
 			  "GPIO20",
 			  "GPIO21",
 			  "GPIO22",
-			  "GPIO23",
-			  "GPIO24",
+			  "MM_WAKE",
+			  "MM_BUSY",
 			  "GPIO25",
--- a/arch/arm/boot/dts/broadcom/bcm2711-rpi-4-b.dts
+++ b/arch/arm/boot/dts/broadcom/bcm2711-rpi-4-b.dts
@@ -97,9 +97,9 @@
 			  "GPIO15",		/* 15 */
 			  "GPIO16",
-			  "GPIO17",
+			  "MM_RESET",
 			  "GPIO18",
 			  "GPIO19",
 			  "GPIO20",		/* 20 */
 			  "GPIO21",
 			  "GPIO22",
-			  "GPIO23",
-			  "GPIO24",
+			  "MM_WAKE",
+			  "MM_BUSY",
 			  "GPIO25",		/* 25 */
--- a/arch/arm/boot/dts/broadcom/bcm2711-rpi-cm4.dts
+++ b/arch/arm/boot/dts/broadcom/bcm2711-rpi-cm4.dts
@@ -335,9 +335,9 @@
 			  "RXD1",
 			  "GPIO16",
-			  "GPIO17",
+			  "MM_RESET",
 			  "GPIO18",
 			  "GPIO19",
 			  "GPIO20",
 			  "GPIO21",
 			  "GPIO22",
-			  "GPIO23",
-			  "GPIO24",
+			  "MM_WAKE",
+			  "MM_BUSY",
 			  "GPIO25",
