From a217970260886e2e6521fcde6409f3270d6a7030 Mon Sep 17 00:00:00 2001
From: Lachlan Hodges <lachlan.hodges@morsemicro.com>
Date: Fri, 28 Mar 2025 10:33:43 +1100
Subject: [PATCH] Implement S1G iwinfo freqlist parsing support

---
 iwinfo.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

Index: rpcd-2024.09.17~9f4b86e7/iwinfo.c
===================================================================
--- rpcd-2024.09.17~9f4b86e7.orig/iwinfo.c
+++ rpcd-2024.09.17~9f4b86e7/iwinfo.c
@@ -701,11 +701,17 @@ rpc_iwinfo_freqlist(struct ubus_context
 			f = (struct iwinfo_freqlist_entry *)&res[i];
 			d = blobmsg_open_table(&buf, NULL);
 
-			band = iwinfo_band2ghz(f->band);
-			if (band > 0)
-				blobmsg_add_u32(&buf, "band", band);
+			if (f->band && f->band == IWINFO_BAND_900) {
+				blobmsg_add_u32(&buf, "band", 900);
+			} else {
+				band = iwinfo_band2ghz(f->band);
+				if (band > 0)
+					blobmsg_add_u32(&buf, "band", band);
+			}
+
 			blobmsg_add_u32(&buf, "channel", f->channel);
 			blobmsg_add_u32(&buf, "mhz", f->mhz);
+			blobmsg_add_u32(&buf, "offset", f->offset);
 			blobmsg_add_u8(&buf, "restricted", f->restricted);
 
 			rpc_iwinfo_add_bit_array("flags", f->flags,
