Index: linux-6.6.110/drivers/mmc/core/quirks.h
===================================================================
--- linux-6.6.110.orig/drivers/mmc/core/quirks.h
+++ linux-6.6.110/drivers/mmc/core/quirks.h
@@ -207,6 +207,10 @@ static const struct mmc_fixup __maybe_un
 			      MMC_QUIRK_LENIENT_FN0 |
 			      MMC_QUIRK_BLKSZ_FOR_BYTE_MODE),
 
+	SDIO_FIXUP_COMPATIBLE("morse,mm6104", add_quirk, MMC_QUIRK_NO_18V),
+
+	SDIO_FIXUP_COMPATIBLE("morse,mm610x", add_quirk, MMC_QUIRK_NO_18V),
+
 	END_FIXUP
 };
 
Index: linux-6.6.110/drivers/mmc/core/sdio.c
===================================================================
--- linux-6.6.110.orig/drivers/mmc/core/sdio.c
+++ linux-6.6.110/drivers/mmc/core/sdio.c
@@ -739,7 +739,7 @@ try_again:
 	 * try to init uhs card. sdio_read_cccr will take over this task
 	 * to make sure which speed mode should work.
 	 */
-	if (rocr & ocr & R4_18V_PRESENT) {
+	if ((rocr & ocr & R4_18V_PRESENT)  && !(card->quirks & MMC_QUIRK_NO_18V)) {
 		err = mmc_set_uhs_voltage(host, ocr_card);
 		if (err == -EAGAIN) {
 			mmc_sdio_pre_init(host, ocr_card, card);
Index: linux-6.6.110/include/linux/mmc/card.h
===================================================================
--- linux-6.6.110.orig/include/linux/mmc/card.h
+++ linux-6.6.110/include/linux/mmc/card.h
@@ -298,6 +298,7 @@ struct mmc_card {
 #define MMC_QUIRK_BROKEN_CACHE_FLUSH	(1<<16)	/* Don't flush cache until the write has occurred */
 #define MMC_QUIRK_BROKEN_SD_POWEROFF_NOTIFY	(1<<17) /* Disable broken SD poweroff notify support */
 #define MMC_QUIRK_NO_UHS_DDR50_TUNING	(1<<18) /* Disable DDR50 tuning */
+#define MMC_QUIRK_NO_18V   (1<<19)     /* Ignore voltage negotiation and don't use 1.8v */
 
 	bool			written_flag;	/* Indicates eMMC has been written since power on */
 	bool			reenable_cmdq;	/* Re-enable Command Queue */
